tools/protoxform/protoprint.py

```diff
@@ -553,12 +568,78 @@ def format_reserved(enum_or_msg_proto):
     return reserved_fields
 
 
+def validate_deprecation_version(version: str):
+    """Validates that a given version string is of format X.Y, where X and Y are
+  integers, X>0, and Y>=0.
+
+  Args:
+    version: a minor version deprecation annotation value (see api/envoy/annotations/deprecation.proto)
+
+  Returns:
+    True iff the given string represents a valid X.Y version.
+  """
+    if version.count('.') != 1:
+        return False
+    # Validate major and minor parts.
+    try:
+        major, minor = [int(x) for x in version.split('.')]
+        if major <= 0:
+            return False
+        if minor < 0:
+            return False
+    except ValueError:
+        return False
+    return True
+
+
 class ProtoFormatVisitor(visitor.Visitor):
     """Visitor to generate a proto representation from a FileDescriptor proto.
 
     See visitor.Visitor for visitor method docs comments.
     """
 
+    def __init__(self, api_version_file_path, frozen_proto):
+        current_api_version = api_version_utils.get_api_version(api_version_file_path)
+        self._deprecated_annotation_version_value = '{}.{}'.format(current_api_version.major,
+                                                                   current_api_version.minor)
+        self._needs_deprecation_annotation_import = False
+        self._frozen_proto = frozen_proto
+
+    def _add_deprecation_version(self, field_or_evalue, deprecation_tag, disallowed_tag):
+        """Adds a deprecation version annotation if needed to the given field or enum value.
+        The annotation is added if all the following hold:
+        - The field or enum value are marked as deprecated.
+        - The proto is not frozen.
+        - The field or enum value are not marked as hidden.
+        - The field or enum value do not already have a version annotation.
+        - The field or enum value name is not ENVOY_DEPRECATED_UNAVIALABLE_NAME.
+        If a field or enum value are marked with an annotation, the
+        _needs_deprecation_annotation_import flag is set.
+        The function also validates that if a field or enum value already have the deprecated
+        annotation value, then this value is a valid one ("X.Y" where X and Y are valid major,
+        and minor versions, respectively).
+        """
+        if field_or_evalue.options.deprecated and not self._frozen_proto and \
+                not protoxform_options.has_hide_option(field_or_evalue.options):
+            # If the field or enum value has annotation from deprecation.proto, need to import it.
+            if field_or_evalue.options.HasExtension(deprecation_tag) or \
+                    field_or_evalue.options.HasExtension(disallowed_tag):
```

Comment:
This would then set/reset the flag on each invocation (not a flag that is only raised once).
I can do the following:
```
self._needs_deprecation_annotation_import = (
    self._needs_deprecation_annotation_import
    or field_or_evalue.options.HasExtension(deprecation_tag)
    or field_or_evalue.options.HasExtension(disallowed_tag))
```

Do you think it's more understandable this way?

Reply:
yep wfm